evap/contributor/tests/test_forms.py

```diff
@@ -65,6 +65,36 @@ def test_managers_only(self):
         self.assertEqual(expected, set(formset.forms[0].fields['questionnaires'].queryset))
         self.assertEqual(expected, set(formset.forms[1].fields['questionnaires'].queryset))
 
+    def test_locked_questionnaire(self):
+        """
+            Asserts that locked (general) questionnaires cannot be changed.
+        """
+        questionnaire = baker.make(Questionnaire, type=Questionnaire.Type.TOP, is_locked=False, visibility=Questionnaire.Visibility.EDITORS)
+        locked_questionnaire = baker.make(Questionnaire, type=Questionnaire.Type.TOP, is_locked=True, visibility=Questionnaire.Visibility.EDITORS)
+
+        evaluation = baker.make(Evaluation)
+        evaluation.general_contribution.questionnaires.add(questionnaire)
+
+        form_data = get_form_data_from_instance(EvaluationForm, evaluation)
+        form_data["general_questionnaires"] = [questionnaire.pk, locked_questionnaire.pk]  # add locked questionnaire
+
+        form = EvaluationForm(form_data, instance=evaluation)
+
+        # Assert form is valid, but locked questionnaire is not added
+        form.save()
+        self.assertEqual(1, evaluation.general_contribution.questionnaires.count())
+
+        evaluation.general_contribution.questionnaires.add(locked_questionnaire)
+
+        form_data = get_form_data_from_instance(EvaluationForm, evaluation)
+        form_data["general_questionnaires"] = [questionnaire.pk]  # remove locked questionnaire
+
+        form = EvaluationForm(form_data, instance=evaluation)
+
+        # Assert form is valid, but locked questionnaire is not removed
+        form.save()
+        self.assertEqual(2, evaluation.general_contribution.questionnaires.count())
```

Comment:
Have you considered asserting that the correct objects are in there? We wouldn't need to spend a though on whether the existing object could have been added again or the wrong one was removed / whatever:
```python
self.assertEqual({questionnaire, locked_questionnaire}, set(evaluation.general_contribution.questionnaires))
```

(There are 4 assertEqual statements in this PR where this could be done)

Reply:
Good idea! I modified this as suggested.